### PR TITLE
Python Metaclasses for Metaprogramming

### LINKED ISSUE #4692: Python Metaclasses for Metaprogramming
#### Proposed title of the article
[Languages] Python Metaclasses for Metaprogramming

#### Proposed article introduction
As strange as it may seem, if you've ever dealt with decorators or metaclasses, you've already done some kind of metaprogramming. Metaprogramming, in a nutshell, is code that manipulates other code.
We'll talk about Metaclasses in this article, including when and why to use them and what to do instead.

#### Key takeaways 
Before beginning this tutorial, students should review "Decorators in Python" and "OOPS principles in Python."
By the end of this tutorial, the reader will be able to understand:
- What are metaclasses
- Cases where the users can use the metaclass
- What it entails in customizing the metaclass
- How to use a metaclass to fix issues even those involving complicated programming

#### Article quality
I will provide an in-depth explanation of the concept as well as incorporate a language easy to be understood by the reader. The reader will appreciate the fact that this tutorial shall provide practicable examples, therefore, making it beginner-friendly.

#### References
N/A

### PR DESCRIPTION
**Pull request by Dick7603**
#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [ ] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

#### Topic suggestion this closes
`This closes #4692

### CODE REVIEW

**Comment by Kayere** [comment]
Hello @Dick7603,

Upon running your article through our 3rd party plagiarism checker it seemed to raise a few flags and the % was higher than we typically accept.

Please see attached PDF - and revisit the article to ensure we are contributing wholly unique and original content.

Be sure to see our resources page to see more info on plagiarism and what is considered as such.

[Python Metaclasses for Metaprogramming #5345 Copyleaks Report.pdf](https://github.com/section-engineering-education/engineering-education/files/7692354/Python.Metaclasses.for.Metaprogramming.5345.Copyleaks.Report.pdf)

File: content/articles/python-metaclasses-for-metaprogramming/index.md [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+```
+
+We may proceed by starting with a class that does not utilize any methods or variables from the class. Then, the method variables and class methods will be specified in the following steps. In the end, we will create the final object.
+
+### Reasons to use metaclasses instead of functions
+- It is apparent what the goal is.  In `UpperAttrMetaclass(type)`, it is clear what comes next.
+- OOP is an option. Metaclasses may inherit the parent method.
+- A class's subclasses are instances of its metaclass if the metaclass-class is given, but not the metaclass function.
+- You may improve the organization of your code. Using metaclasses for a simple example like this is a waste of time. It is usually reserved for a more difficult task. Making several methods and grouping them into a single class makes the code more readable.
+- It is possible to hook on `new`, `init`, and `call` hooks. So you may do other things, as a result.
```
Comment (jerimkaura) [inline content/articles/python-metaclasses-for-metaprogramming/index.md:95]: Explain these reasons in detail. The one-line statements provided are not clear enough.

File: content/articles/python-metaclasses-for-metaprogramming/index.md [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+### Instances to use metaclasses
+The general public seldom utilizes metaclasses since they are primarily employed in complex scenarios. However, metaclasses may be used in the following limited number of situations.
+- While generating down the hierarchies for inheritance.
+- If the user wishes to alter the class automatically.
+- If the user is a developer of application programming interfaces.
```
Comment (jerimkaura) [inline content/articles/python-metaclasses-for-metaprogramming/index.md:101]: Elaborate on these to add more value.

File: content/articles/python-metaclasses-for-metaprogramming/index.md [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    def mul(own, x, y):  
+        return x*y  
+    def div(own, x, y):  
+        return x/y 
+       
+enjoyer_cal = Calc()  
+print(enjoyer_cal.add(6, 4))  
+print(enjoyer_cal.mul(8, 9))  
+print(enjoyer_cal.div(30, 5)) 
+```
```
Comment (jerimkaura) [inline content/articles/python-metaclasses-for-metaprogramming/index.md:175]: Explain the code snippet here well. Tell the user what it does and how it does it.

**Review by jerimkaura — CHANGES_REQUESTED** [review]

**Review by jerimkaura — CHANGES_REQUESTED** [review]
Hello @Dick7603 

You have an amazing article here. However, please go over the article to work on the changes highlighted and to ensure readability. You need to ensure every article is clear to the readers. 

Work on the changes then ping me here or via slack.

**Comment by jerimkaura** [comment]
Hello @Dick7603 

Please work on all the changes so that I can proceed with the review. The commit did not take in all the suggested changes.

**Review by jerimkaura — CHANGES_REQUESTED** [review]
Hello @Dick7603 

Please work on all the changes so that I can proceed with the review. The commit did not take in all the suggested changes.

**Review by jerimkaura — APPROVED** [review]
Hello @Dick7603 

Great piece here :rocket: Take note of the changes for better articles in the future. Great Work!

File: content/articles/python-metaclasses-for-metaprogramming/index.md [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+- [What are metaclasses](#what-are-metaclasses)
+- [Understanding metaclasses with examples](#understanding-metaclasses-with-examples)
+- [Reasons to use metaclasses instead of functions](#reasons-to-use-metaclasses-instead-of-functions)
+- [Instances to use metaclasses](#instances-to-use-metaclasses)
+- [Customizing metaclasses](#customizing-metaclasses)
+- [Using a metaclass to fix an issue](#using-a-metaclass-to-fix-an-issue)
+- [Conclusion](#conclusion)
+
+### What are metaclasses?
+Metaclasses are classes whose instances are other classes used to build objects. Metaclasses can be created by modifying how classes are instantiated.
```
Comment (WanjaMIKE) [inline content/articles/python-metaclasses-for-metaprogramming/index.md:36]: This statement appears unclear. Could you rephrase it.

File: content/articles/python-metaclasses-for-metaprogramming/index.md [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+Town_object = Town()  
+print("Sort of Town_object is:", type(Town_object))
+```
+
+#### Output
+```python
+Sort of Town_object is: <class '__main__.Town'>
+```
+
+A class is an instance of metaclass since it is an object in Python. Therefore, various class types can be found via the metaclass. This, therefore, establishes new classes and class objects. For instance, `town's` type can be found as below:
```
Comment (WanjaMIKE) [inline content/articles/python-metaclasses-for-metaprogramming/index.md:77]: Rephrase this paragraph

File: content/articles/python-metaclasses-for-metaprogramming/index.md [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+### Reasons to use metaclasses instead of functions
+Although we have equally useful functions, we need metaclasses for the following reasons.
+- The aim of metaclasses is precise. At the point when you read `UpperAttrMetaclass(type)`, you know what will follow
+- You can utilize OOP. Metaclasses may inherit the parent method.
+- A class's subclasses are instances of its metaclass if the metaclass-class is given, but not the metaclass function.
+- You may improve the organization of your code. Using metaclasses for a simple example like this is a waste of time. It is usually reserved for a more difficult task. Making several methods and grouping them into a single class makes the code more readable.
+- It is possible to hook on `new`, `init`, and `call` hooks. 
+
+### Instances to use metaclasses
+The general public seldom utilizes metaclasses since they are primarily employed in complex scenarios. However, metaclasses may be used in the following limited number of situations:
```
Comment (WanjaMIKE) [inline content/articles/python-metaclasses-for-metaprogramming/index.md:121]: Replace `the general public` with developers to be more specific.

File: content/articles/python-metaclasses-for-metaprogramming/index.md [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+### Customizing the Metaclass
+The user-created metaclass must inherit the type metaclass and, in most cases, override it, such as:
+
+- `__new__()`: This method comes before `__int__()` in the class hierarchy. With this method, we can create and return the object. Remember that we can override it to control how objects are created.
+
+- `__int__()`: This function enables the initialization of the objects that have been created, and therefore, to call the object, you have to pass it as an argument
+
+The `type()` function may be used directly by the user to construct classes, and there are possibilities of utilizing this function in different ways.
+
+The preceding example shows that the type may be sent in as a single parameter. It may be called with three different arguments by the user. Creating the class will be done by this. Some of the arguments include:
```
Comment (WanjaMIKE) [inline content/articles/python-metaclasses-for-metaprogramming/index.md:135]: This statement is unclear `Creating the class will be done by this?`

**Comment by Dick7603** [comment]
hello @WanjaMIKE, I have implemented the changes your requested please check

File: content/articles/python-metaclasses-for-metaprogramming/index.md [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+- [What are metaclasses](#what-are-metaclasses)
+- [Understanding metaclasses with examples](#understanding-metaclasses-with-examples)
+- [Reasons to use metaclasses instead of functions](#reasons-to-use-metaclasses-instead-of-functions)
+- [Instances to use metaclasses](#instances-to-use-metaclasses)
+- [Customizing metaclasses](#customizing-metaclasses)
+- [Using a metaclass to fix an issue](#using-a-metaclass-to-fix-an-issue)
+- [Conclusion](#conclusion)
+
+### What is a metaclass?
+A metaclass in Python is a class of a class that defines how a class behaves. A class is itself an instance of a metaclass. A class in Python defines how the instance of the class will behave.
```
Comment (WanjaMIKE) [inline content/articles/python-metaclasses-for-metaprogramming/index.md:36]: You copied this content directly from https://www.datacamp.com/community/tutorials/python-metaclasses

copycat

**Comment by WanjaMIKE** [comment]
I am closing this PR because it seems it wasn't publish-ready. You can make the changes outside the queue and create a new duplicate PR.
